tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -2319,20 +2324,137 @@ private[optimizer] abstract class OptimizerCore(
     }
   }
 
+  private def constantFoldBinaryOp_except_String_+(op: BinaryOp.Code,
+      lhs: Literal, rhs: Literal)(implicit pos: Position): Literal = {
+    import BinaryOp._
+
+    @inline def int(lit: Literal): Int = (lit: @unchecked) match {
+      case IntLiteral(value) => value
+    }
+
+    @inline def long(lit: Literal): Long = (lit: @unchecked) match {
+      case LongLiteral(value) => value
+    }
+
+    @inline def float(lit: Literal): Float = (lit: @unchecked) match {
+      case FloatLiteral(value) => value
+      case IntLiteral(value)   => value.toFloat
```

Comment:
That said, I now realize there is another problem: float operations are not supposed to accept int values, because `int` is not a subtype of `float`. So in fact this `case` should not be there.